adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -0,0 +1,636 @@
+                                            return Future.succeededFuture();
+                                        }).otherwise(registrationAssertionFailure -> {
+                                            LOG.error("Error asserting device registration. Sending error response to "
+                                                    + "application with code [{}]", HTTP_FORBIDDEN,
+                                                    registrationAssertionFailure);
+                                            TracingHelper.logError(receivedCommandContext.getCurrentSpan(),
+                                                    registrationAssertionFailure);
+                                            sendResponseToApplication(command, loraDeviceId,
+                                                    getHttpResponseWithCode(HTTP_FORBIDDEN,
+                                                            registrationAssertionFailure.getMessage()),
+                                                    receivedCommandContext.getCurrentSpan().context());
+                                            return null;
+                                        }))
+                        .otherwise(loraGatewayException -> {
+                            LOG.error("Error getting lora device gateway. Sending error response to application with "
+                                    + "code [{}]", HTTP_INTERNAL_ERROR, loraGatewayException);
+                            TracingHelper.logError(receivedCommandContext.getCurrentSpan(), loraGatewayException);
+                            sendResponseToApplication(command, loraDeviceId,
+                                    getHttpResponseWithCode(HTTP_INTERNAL_ERROR,
+                                            loraGatewayException.getMessage()),
+                                    receivedCommandContext.getCurrentSpan().context());
+                            return null;
+                        });
+
+                return Future.succeededFuture();
+            } else {
+                LOG.debug("Got invalid command {} for actual lora device '{}'", command,
+                        command.getApplicationProperties().get(MessageHelper.APP_PROPERTY_DEVICE_ID));
+                return Future.failedFuture("Malformed command message.");
+            }
+        }).otherwise(validationException -> {
+            LOG.error("Error trying to send command '{}'", command, validationException);
+            TracingHelper.logError(receivedCommandContext.getCurrentSpan(), validationException);
+            receivedCommandContext
+                    .reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, validationException.getMessage()), 1);
+            return null;
+        });
+    }
+
+    private static boolean isValidLoraCommand(final Command command) {
+        try {
+            final String payload = command.getPayload().toJsonObject()
+                    .getString(LoraConstants.FIELD_LORA_DOWNLINK_PAYLOAD);
```

Comment:
FMPOV the command's payload *is* the payload to be sent. Why would it be encapsulated in a JSON String property?